### PR TITLE
feat: support start_key in v3 readdir api

### PR DESCRIPTION
**Pull request by wfnuser**
To solve https://github.com/apache/apisix/issues/2853#issuecomment-741506186

The solution I proposed is that we use start_key and limit to limit the range of readdir.
And we can learn whether we should continue read or not from the result.

When we need to pagination, we just use a loop with the constant limit and the changeable start_key to read the whole dir.

### CODE REVIEW

File: t/v3/dir.t [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+            local res, err = etcd:set("/dir", "test")
+            check_res(res, err)
+            local res, err = etcd:set("/dir/a", "a")
+            check_res(res, err)
+            local res, err = etcd:set("/dir/b", "b")
+            check_res(res, err)
+            local res, err = etcd:set("/dir/c", "c")
+            check_res(res, err)
+
+            res, err = etcd:readdir("/dir", {limit = 3})
```
Comment (spacewander) [inline t/v3/dir.t:83]: Would be better to use another number than 3, as the whole number of keys is 3

File: t/v3/dir.t [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+            if tab_nkeys(res.body.kvs) == 3 then
+                ngx.say("ok")
+            else
+                ngx.say("failed")
+            end
+
+            res, err = etcd:readdir("/dir", {start_key = "/dir/a"})
```
Comment (spacewander) [inline t/v3/dir.t:90]: Would be better to use `/dir/b`, since `/dir/a` is already the first key.

**Review by spacewander — CHANGES_REQUESTED** [review]

File: api_v3.md [hunk 3]
```diff
@@ -192,7 +192,8 @@ res = cli:watchcancel(http_cli)
     * `sort_order`: (int [SortNone:0, SortAscend:1, SortDescend:2]) sort_order is the order for returned sorted results.  
     * `sort_target`: (int [SortByKey:0, SortByVersion:1, SortByCreateRevision:2, SortByModRevision:3, SortByValue:4]) sort_target is the key-value field to use for sorting.
     * `keys_only`: (bool) keys_only when set returns only the keys and not the values.  
-    * `count_only`: (bool) count_only when set returns only the count of the keys in the range. 
+    * `count_only`: (bool) count_only when set returns only the count of the keys in the range.
+    * `start_key`: (string value) the start_key is the start key for the dir range.
```
[the remark this answers is not in the document]
Comment (spacewander) [reply api_v3.md:196]: Just a suggestion. If you think `start_key` is better, we can keep the name.

File: t/v3/dir.t (hunk 1 above)
Comment (wfnuser) [reply t/v3/dir.t:83]: Not really. The whole number of keys is actually 4.

Comment (wfnuser) [reply t/v3/dir.t:83]: Since the /dir is the first key.

File: api_v3.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (wfnuser) [reply api_v3.md:196]: I agree that range_start is better.

File: t/v3/dir.t (hunk 2 above)
Comment (wfnuser) [reply t/v3/dir.t:90]: /dir is the first key. But I also agree with that /dir/a is better.

File: t/v3/dir.t (hunk 1 above)
Comment (spacewander) [reply t/v3/dir.t:83]: Look like we need to update the doc to clarify that `limit` also apply on the dir key.

File: t/v3/dir.t (hunk 2 above)
Comment (spacewander) [reply t/v3/dir.t]: There is a problem, I remember APISIX requires the first key returned by `readdir` is the dir key. 
@nic-chen @Yiyiyimu 
Please correct me if I am wrong.

Comment (Yiyiyimu) [reply t/v3/dir.t]: @spacewander Yes I think `/dir` would be the first key, if `/dir` does not contains a k-v pair and just the directory, it just would have no value. See [get_format](https://github.com/apache/apisix/blob/master/apisix/core/etcd.lua#L108-L119)

**Comment by Yiyiyimu** [comment]
Hi @wfnuser do you have any progress?

**Comment by wfnuser** [comment]
> Hi @wfnuser do you have any progress?

I'm still waiting for the merge. Haha. Shall we merge this request now?

**Comment by Yiyiyimu** [comment]
> I'm still waiting for the merge. Haha. Shall we merge this request now?

The CI haven't passed yet 🤣 Could you have a look at it

**Comment by wfnuser** [comment]
> > I'm still waiting for the merge. Haha. Shall we merge this request now?
> 
> The CI haven't passed yet 🤣 Could you have a look at it

Ok. I thought it passed before...🤣 Let me check it later.

File: t/v3/dir.t (hunk 2 above)
Comment (spacewander) [reply t/v3/dir.t]: @wfnuser
I think you should at least mention this in the doc.

File: t/v3/dir.t (hunk 1 above)
Comment (spacewander) [reply t/v3/dir.t:83]: Need to update the doc.

**Comment by membphis** [comment]
any news? ping @wfnuser

**Comment by wfnuser** [comment]
> any news? ping @wfnuser

Aha... I forgot this pr... I will check it tonight.

**Comment by wfnuser** [comment]
> any news? ping @wfnuser

I'm try to recap the environment of the pr tonight... It won't work now. 
And I found that the learning curve of Nginx:Test really steep, expecially for newbee to run the test.
Are there any way easier to run the test. Like we can run tests in an encapsulated environment by docker?

I forgot how to set up the environment.. So still need some time, I will try again tomorrow.

The problem I met looks like... It seems I need to install resty.http? But I don't think the path for my resty looks like `lib/?.lua;/usr/local/share/lua/5.3/?.lua;/usr/share/lua/5.1/?.lua;;`

```
#   Failed test 'TEST 1: set + readdir + readdir with range_start + rmdir - pattern "[error]" should not match any line in error.log but matches line "2021/07/06 00:07:00 [error] 26740\#8091418: *1 lua entry thread aborted: runtime error: lib/resty/etcd/v2.lua:2: module 'resty.http' not found:" (req 0)
#       no field package.preload['resty.http']
#       no file 'lib/resty/http.lua'
#       no file '/usr/local/share/lua/5.3/resty/http.lua'
#       no file '/usr/share/lua/5.1/resty/http.lua'
#       no file '/usr/local/Cellar/openresty/1.19.3.2_1/site/lualib/resty/http.ljbc'
#       no file '/usr/local/Cellar/openresty/1.19.3.2_1/site/lualib/resty/http/init.ljbc'
#       no file '/usr/local/Cellar/openresty/1.19.3.2_1/lualib/resty/http.ljbc'
#       no file '/usr/local/Cellar/openresty/1.19.3.2_1/lualib/resty/http/init.ljbc'
#       no file '/usr/local/Cellar/openresty/1.19.3.2_1/site/lualib/resty/http.lua'
```

I will try to figure it out tomorrow. But any help will be greatly appreciate.

**Comment by Yiyiyimu** [comment]
@wfnuser If that's the only lib missing, maybe you could directly copy your lib to the desired path 🤣

**Comment by wfnuser** [comment]
> @wfnuser If that's the only lib missing, maybe you could directly copy your lib to the desired path 🤣

ok... Is it the usual solution?

**Comment by wfnuser** [comment]
> @wfnuser If that's the only lib missing, maybe you could directly copy your lib to the desired path 🤣

Wow. You study at Georgia Tech? That's cool!

**Comment by wfnuser** [comment]
@Yiyiyimu Maybe we can approve to run the workflows.

**Comment by wfnuser** [comment]
@membphis Hi. Maybe you need to run the workflow again.

**Comment by Yiyiyimu** [comment]
> ok... Is it the usual solution?

Of course not it's just a lazy solution 🤣 And thanks for your flattering 😝

**Comment by wfnuser** [comment]
> > ok... Is it the usual solution?
> 
> Of course not it's just a lazy solution 🤣 And thanks for your flattering 😝

Since all checks are passed. Shall we merge this pr now?

**Comment by wfnuser** [comment]
@membphis  Shall we merge it?

**Comment by wfnuser** [comment]
Anyone?

**Review by Yiyiyimu — APPROVED** [review]

File: t/v3/dir.t [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+            res, err = etcd:readdir("/dir", {limit = 3})
+            if tab_nkeys(res.body.kvs) == 3 then
+                ngx.say("ok")
+            else
+                ngx.say("failed")
+            end
+
+            res, err = etcd:readdir("/dir", {range_start = "/dir/b"})
+            check_res(res, err, "b")
```
Comment (nic-chen) [inline t/v3/dir.t:91]: we could add a test case about "a" not in response

Comment (wfnuser) [reply t/v3/dir.t:91]: I checked the result of etcd looks exactly as "/dir/b" and "/dir/c". Is that what you need?

**Comment by wfnuser** [comment]
I have already update the test. @nic-chen

File: t/v3/dir.t [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+run_tests();
+
+__DATA__
+
+=== TEST 1: set + readdir + readdir with range_start + rmdir
+--- http_config eval: $::HttpConfig
+--- config
+    location /t {
+        content_by_lua_block {
+            local tab_nkeys     = require "table.nkeys"
```
Comment (membphis) [inline t/v3/dir.t:67]: bad style.

```lua
local tab_nkeys = require("table.nkeys")
```

File: t/v3/dir.t [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+            local etcd, err = require "resty.etcd" .new({protocol = "v3"})
```
Comment (membphis) [inline t/v3/dir.t:68]: ditto

File: t/v3/dir.t (hunk 5 above)
Comment (wfnuser) [reply t/v3/dir.t]: done

File: t/v3/dir.t (hunk 6 above)
Comment (wfnuser) [reply t/v3/dir.t]: done

**Review by nic-chen — APPROVED** [review]

**Review by spacewander — CHANGES_REQUESTED** [review]
I think we should use a new API to do the range filter. As people have pointed out: https://github.com/api7/lua-resty-etcd/pull/107#discussion_r544059582. We can't use it in readdir.

**Comment by wfnuser** [comment]
> I think we should use a new API to do the range filter. As people have pointed out: [#107 (comment)](https://github.com/api7/lua-resty-etcd/pull/107#discussion_r544059582). We can't use it in readdir.

You mean we should remove the first key, since it is just the parent folder? Can you make the idea more clear.

**Comment by spacewander** [comment]
A new, separate range API is more clear.

**Review by membphis — DISMISSED** [review]

**Comment by wfnuser** [comment]
> A new, separate range API is more clear.

Ok. So the `readdir` api should not support `range_start`?

**Comment by wfnuser** [comment]
> > A new, separate range API is more clear.
> 
> Ok. So the `readdir` api should not support `range_start`?

If it should, I think we will support range_start in readdir and add another api. I don't see the reason why not range_start should not be support in readdir?

**Comment by wfnuser** [comment]
Suit yourself.

**Pull request closed without merging**